sycl/doc/DeviceGlobal.md

```diff
@@ -110,60 +101,137 @@ However, when the property is present, it has a member variable which is the
+can be declared.  Rather than have the front-end recognize the name of the
+`device_global` type, the front-end uses this attribute to know which
+restrictions to enforce for this type.
+
+**NOTE**: The front-end does currently recognize the `specialization_id` class
+by its name, and it has hard-coded knowledge that variables declared with this
+type have certain restrictions.  If we wanted to avoid having the front-end
+recognize the `specialization_id` class by its name, we could decorate that
+class with `[[__sycl_detail__::restrictions(spec_id)]]`.  Note that we would
+need a different parameter name because the restrictions for
+`specialization_id` are different from those for `device_global`.
+
+#### Declaration of member functions to copy device global variables
+
 The headers are also updated to add the new `copy()` and `memcpy()` member
 functions to `handler` and `queue` which copy data to or from a device global
 variable.  These declarations use SFINAE such that they are conditionally
 available depending on the `host_access` property.
 
-### New LLVM IR attributes
-
-Two new attributes are added to communicate information about device global
-variable to the `sycl-post-link` tool: `sycl-unique-id` and
-`sycl-device-global-image-life`.  As described below, the device compiler
-front-end is responsible for adding the attributes to the LLVM IR.
-
-Each device global variable is decorated with `sycl-unique-id`, which provides
-a unique string identifier for each device global variable.  This string will
-also be used to name the variable in SPIR-V, so it's better for debuggability
-if the string matches the mangled name for variables with external linkage.
-This is not possible, though, for variables with internal linkage because the
-mangled name is not unique in this case.  For these variables, we use the
-mangled name and append a unique suffix.
+### Changes to the DPC++ front-end
 
-Each device global variable that has the `device_image_scope` property is also
-decorated with the `sycl-device-global-image-life` attribute.
+There are several changes to the device compiler front-end:
 
-Note that language rules ensure that `device_global` variables are always
-declared at namespace scope (i.e. a global variable), and LLVM IR [allows
-attributes to be attached to global variables][2].
+* The front-end adds a new LLVM IR attribute `sycl-unique-id` to the definition
+  of each `device_global` variable, which provides a unique string identifier
+  for each device global variable.  The rules for creating this string are the
+  same as `__builtin_sycl_unique_stable_id`, so the front-end can use the same
+  algorithm when generating the string.
 
-[2]: <https://llvm.org/docs/LangRef.html#global-attributes>
+* The front-end checks for restrictions on variable declarations using the
+  `device_global` type.  As described above, the front-end uses the
+  `[[__sycl_detail__::restrictions()]]` attribute (rather than the class name)
```

Comment:
Minor question.
`[[__sycl_detail__::restrictions()]]` attribute purpose/spirit is to add restrictions on top of regular SYCL restrictions or it can relax some of them as well or redefine completely?
If second or third, then `[[__sycl_detail__::global_variable_allowed]]` is not really required, but probably OK for symmetry.

Reply:
It's purpose is to implement the restrictions which are specified in the core SYCL spec or in extension specs.  Therefore, it is neither adding new restrictions nor relaxing them.

There are two cases in SYCL where we have special restrictions that are tied to variables of a specific type.  Variables of type `specialization_id` have a special set of restrictions, and now variables of type `device_global` have a different set of special restrictions.  In both cases, the front-end will need to have hard-coded knowledge about how to check these restrictions and diagnose an error if the application violates them.

How will the front-end identify these variables, though?  One way would be for the front-end to recognize the name of the variable's type, which is what we do today for `specialization_id`.  We could do that also for `device_global`, but that seems messy to me.  For example, if we change the namespace of `device_global` from `sycl::ext::oneapi` to just `sycl`, then we would need to change the front-end to recognize the new namespace.

It seemed cleaner to me to communicate the restrictions via an attribute, rather than having the front-end recognize the name of the type.  This is the only reason we have `[[__sycl_detail__::restrictions()]]`.  If the front-end sees a variable whose type has the attribute `[[__sycl_detail__::restrictions(device_global)]]`, then it knows to apply the special restrictions for device global variables, and this logic will continue to work regardless of the exact spelling of the type's name.

I think you are suggesting that we somehow combine `[[__sycl_detail__::restrictions()]]` with `[[__sycl_detail__::global_variable_allowed]]`.  I think that would be possible, but we should choose a different name for the attribute.  For example, we could name it `[[__sycl_detail__::type_rules()]]`, and then allow something like:

```
[[__sycl_detail__::type_rules(device_global, allow_global_variable)]]
```

Since this mostly affects the front-end, I think we should allow the CFE team to decide which is easier for them.  Tagging @elizabethandrews and @premanandrao.

Reply:
My idea was to use `[[__sycl_detail__::global_variable_allowed]]` to identify any global variable that can legally be referenced from device code.  If we invent some new extension in the future that adds another type that is allowed as a global variable, then we can also decorate that new type with `[[__sycl_detail__::global_variable_allowed]]`.  I was hoping that by adding a general mechanism like this we can avoid asking the CFE team for support for new features in the future.

Therefore, the `[[__sycl_detail__::global_variable_allowed]]` attribute does not necessarily mean that the variable is a "device global".  It just means that the type is allowed to be used to declare a global variable that is referenced in device code.

Reply:
Oh I see. Presumably though, any new extension that adds a type which can be used as a global variable will have some new (or same) set of restrictions. For the former, FE intervention will be required to add the restrictions. For the latter, a single attribute will suffice. 

Another concern here is that in `[[__sycl_detail__::restrictions(device_global)]]`, `device_global` is the name of a type we haven't parsed yet. When parsing the attribute arguments, the class isn't declared yet. So to make this work, we'd have to late parse the attribute. I'm not very familiar with what this involves because I've never worked on an attribute that required it. I spoke briefly with Aaron about it and he mentioned its never been done for attributes which apply to struct types before but we could potentially explore it. 

IMO, having an attribute which defines a certain behavior makes the most sense i.e. a single attribute which says all types with _xyz_ attribute has _abc_ behavior

@premanandrao  @gmlueck thoughts?

Reply:
> Presumably though, any new extension that adds a type which can be used as a global variable will have some new (or same) set of restrictions.

Not necessarily.  For example, the implementation of [device-side-asserts](https://github.com/intel/llvm/pull/3767) added a special global variable.  The front-end needs to avoid issuing an error for kernels that reference this global variable, but there are no other special restrictions that the front-end needs to check.  There could be other cases like this in the future where `[[__sycl_detail__::global_variable_allowed]]` would be useful.  We could even consider changing the implementation of device-side-asserts to use `[[__sycl_detail__::global_variable_allowed]]`.

> Another concern here is that in [[__sycl_detail__::restrictions(device_global)]], device_global is the name of a type we haven't parsed yet.

This is not really the case.  Don't think of the parameter name `device_global` as the name of a type.  Instead, think of it as a name that the front-end intrinsically knows, which just happens to have the same spelling as a type.  This is similar to other clang attributes like [`__attribute__((objc_method_family(X)))`](https://clang.llvm.org/docs/AttributeReference.html#objc-method-family) where the `X` parameter is one of several possible names that the compiler intrinsically knows.

> IMO, having an attribute which defines a certain behavior makes the most sense i.e. a single attribute which says all types with xyz attribute has abc behavior where abc is the allowances/restrictions/etc

I think you are saying that we should not add `[[__sycl_detail__::global_variable_allowed]]` and instead use just `[[__sycl_detail__::restrictions(device_global)]]`.  When the front-end sees that attribute, it will do two things:

* Avoid diagnosing an error if a global variable of that type is referenced from device code.
* Impose certain restrictions that are specific to the `device_global` type.

That will certainly work, and I can change the design this way if you prefer.  I think the downside is that we will need to add a new attribute in the future if there is some case where a type has no special restrictions aside from being allowed as the type of a global variable.  Let me know what you think.

Reply:
I favor the two attributes approach, but it is not a very strong preference.  @elizabethandrews has spent far more time on this, so if she has a preference, I would defer to hers.

Reply:
> I'm new to feature/language design, so if I'm being honest, I do not know the right answer here. From an implementation stand-point, a single attribute which takes no argument is obviously easier but from a language design perspective I do not know.

These are all implementation details, not language design.  None of these `__sycl_detail__` attributes will be documented as part of the SYCL/DPC++ language.

FWIW, I also think the two attribute approach is better, but it is not a very strong preference.  I suspect there will be cases in the future where we have other global variable types that can be referenced from device code.  If we have a dedicated attribute `[[__sycl_detail::global_variable_allowed]]`, there will be a single test in the CFE like this:

```
if (!/* type has "global_variable_allowed" */) {
  // issue diagnostic
}
```

Without the dedicated attribute, the CFE will need to have an ever-growing list of checks:

```
if (!/*type has "device_global"*/ && !/*type has "other thing*/ && !/*etc*/) {
  // issue diagnostic
}
```

> If the answer is 'no FE intervention will be required if we reuse attributes', that is not true for [[__sycl_detail__::restrictions(device_global)]]

You are correct that it is not true for "device_global" because that type requires other restrictions that the CFE must implement.  I suspect there could be other types, though, that do not have any other restrictions.  In such a case, having a dedicated `[[__sycl_detail::global_variable_allowed]]` could avoid CFE changes in some cases.

It's hard to predict the future, though, so it's hard to say for sure if having a dedicated `[[__sycl_detail::global_variable_allowed]]` will save future effort.  To me, adding this second attribute seemed like a small investment that could be helpful in the future.  If you prefer having a single `[[__sycl_detail::device_global]]` attribute, though, I don't mind changing the design.

Reply:
My concerns had more to do with `[[__sycl_detail__::restrictions()]]` than with the 2 attribute approach. In order to implement this, we will need to hardcode the 'types' (i.e. string representing types) accepted by this attribute in the attribute definition. What is the design rationale for going with a general-purpose attribute that requires the compiler to be updated every time a new semantic restriction is desired instead of going with a specific-purpose attribute for the restriction? The compiler FE needs to be updated either way, and it seems likely that someone will find a reason to want to parameterize a restriction at some point in the future, so it seems like there are technical benefits to using specific attributes rather than a single general-purpose one.

IMO what makes most sense here is `[[__sycl_detail::global_variable_allowed]]` and `[[__sycl_detail::device_global]]`. The former for behavior that works across types/future extensions and the latter for device_global restrictions.

Reply:
That works for me.  I'll change the spec to have two attributes with the names you propose.  Just to make sure we're on the same page, the `class device_global` type will use both attributes.  The attribute `[[__sycl_detail::global_variable_allowed]]` causes the compiler to avoid diagnosing an error when there are references to a global variable of this type from device code.  Whereas, the attribute `[[__sycl_detail__::device_global]]` causes the compiler to diagnose an error if a variable of this type is declared in any way that violates the restrictions described here:

https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc#restrictions-on-creating-device-global-objects

Correct?

Reply:
I am still unsure if we can detect the shadowing cases, but otherwise, I am okay with the meanings of the attribute spellings.

Reply:
The spec does not require the compiler to diagnose this error, but of course it is better for our users if it could do this.  I think we can wait until implementation time to see if the CFE can issue a diagnostic in this case.

Reply:
> 

Yes that is what I meant. Thanks!

Reply:
I made this change in df56b04a3386a939ac6d703a8426dcd69553df92